Tests/SonarScanner.MSBuild.PreProcessor.Test/SonarWebServiceTest.cs

```diff
@@ -940,26 +941,54 @@ public async Task GetProperties_Old_Forbidden()
         [TestMethod]
         public void GetProperties_63plus_Forbidden()
         {
-            const string serverUrl = "http://localhost";
-            const string projectKey = "my-project";
-
             var downloaderMock = new Mock<IDownloader>();
             downloaderMock
-                .Setup(x => x.Download(new Uri($"{serverUrl}/api/server/version"), false))
+                .Setup(x => x.Download(new Uri($"{ServerUrl}/api/server/version"), false))
                 .Returns(Task.FromResult("6.3.0.0"));
 
             downloaderMock
-                .Setup(x => x.TryDownloadIfExists(new Uri($"{serverUrl}/api/settings/values?component={projectKey}"), true))
+                .Setup(x => x.TryDownloadIfExists(new Uri($"{ServerUrl}/api/settings/values?component={ProjectKey}"), true))
                 .Throws(new HttpRequestException("Forbidden"));
 
-            var service = new SonarWebService(downloaderMock.Object, serverUrl, logger);
+            var service = new SonarWebService(downloaderMock.Object, ServerUrl, logger);
 
-            Action action = () => _ = service.GetProperties(projectKey, null).Result;
+            Action action = () => _ = service.GetProperties(ProjectKey, null).Result;
             action.Should().Throw<HttpRequestException>();
 
             logger.Errors.Should().HaveCount(1);
         }
 
+        [TestMethod]
+        public void DownloadCache_NullArguments()
+        {
+            ws.Invoking(x => x.DownloadCache(null, "branch")).Should().ThrowAsync<ArgumentNullException>();
+            ws.Invoking(x => x.DownloadCache("key", null)).Should().ThrowAsync<ArgumentNullException>();
+        }
+
+        [TestMethod]
+        public async Task DownloadCache_DeserializesMessage()
+        {
+            using var stream = CreateCacheStream(new AnalysisCacheMsg { Map = { { "key", ByteString.CopyFromUtf8("value") } } });
+            var sut = new SonarWebService(MockIDownloader(stream), ServerUrl, logger);
+
+            var result = await sut.DownloadCache(ProjectKey, ProjectBranch);
+
+            result.Map.Count.Should().Be(1);
+            result.Map["key"].ToStringUtf8().Should().Be("value");
+            logger.AssertInfoLogged("Downloading cache. Project key: project-key, branch: project-branch.");
+        }
+
+        private static Stream CreateCacheStream(IMessage message)
+        {
+            Stream stream = new MemoryStream();
+            message.WriteTo(stream);
+            stream.Seek(0, SeekOrigin.Begin);
+            return stream;
+        }
+
+        private static IDownloader MockIDownloader(Stream stream) =>
+            Mock.Of<IDownloader>(x => x.DownloadStream(It.IsAny<Uri>()) == Task.FromResult(stream));
```

Comment:
_Learning:_ How does this work? 
Does it read the expression, and abuse the `==` operator to split it between `Setup` and `Returns`?

Looks hacky. And practical. I like it.
Would && work for defining two methods?

Reply:
> Does it read the expression, and abuse the == operator to split it between Setup and Returns?

I think so, yes.

> Would && work for defining two methods?

Yes, it works. For properties too 😉